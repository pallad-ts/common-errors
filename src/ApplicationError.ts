export class ApplicationError extends Error {
    code?: string;
    constructor(message: string) {
        super(message);
        this.message = message;
        this.name = 'ApplicationError';
    }
}
