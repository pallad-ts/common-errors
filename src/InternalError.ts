export class InternalError extends Error {
    code?: string;
    constructor(message: string, previousError?: Error) {
        super();
        Object.defineProperty(this, 'previousError', {
            enumerable: false,
            configurable: false,
            value: previousError
        });
        this.message = message;
        this.name = 'InternalError';
        Object.freeze(this);
    }
}
