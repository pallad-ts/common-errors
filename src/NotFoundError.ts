export class NotFoundError<T = { [key: string]: any }> extends Error {
    constructor(message: string, readonly references?: T) {
        super(message);
        this.message = message;
        this.name = 'NotFoundError';
        Object.freeze(this);
    }

    static entity(name: string, reference: { [key: string]: any }) {
        const referenceString = Object.keys(reference)
            .reduce((result, key) => {
                return result.concat([`${key}: ${reference[key]}`]);
            }, [] as string[])
            .join(', ');
        new NotFoundError(`${name} not found${referenceString ? ` ${referenceString}` : ''}`)
    }
}