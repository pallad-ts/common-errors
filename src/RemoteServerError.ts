export class RemoteServerError extends Error {
    code?: string;
    constructor(message: string) {
        super();
        this.message = message;
        this.name = 'RemoteServerError';
    }
}
