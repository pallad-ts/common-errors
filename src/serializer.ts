import {DataNormalizer, Serializer} from "alpha-serializer";
import {NotFoundError} from "./NotFoundError";
import {ApplicationError} from "./ApplicationError";
import {InternalError} from "./InternalError";
import {RemoteServerError} from "./RemoteServerError";
import {TimeoutError} from './TimeoutError';

const prohibitedProperties = ['message', 'stack'];

export function normalizeError(err: Error & { [key: string]: any }) {
	return Object.getOwnPropertyNames(err)
		.filter(propertyName => prohibitedProperties.indexOf(propertyName) === -1)
		.reduce((error: any, propertyName) => {
			if (Object.prototype.propertyIsEnumerable.call(err, propertyName)) {
				error[propertyName] = err[propertyName];
			}
			return error;
		}, {
			message: err.message,
			name: err.name
		});
}

const STANDARD_EXCLUDED_PROPERTIES = ['message', 'name'];

function assignExtraProperties(error: any, currentValue: any, excludedProperties = STANDARD_EXCLUDED_PROPERTIES) {
	for (const [key, value] of Object.entries(currentValue)) {
		if (excludedProperties.includes(key)) {
			continue;
		}
		error[key] = value;
	}
	return error;
}

export function setupSerializer(serializer: Serializer) {
	setup(serializer.normalizer, [
		['Error/NotFound', NotFoundError, value => {
			return assignExtraProperties(
				new NotFoundError(value.message, value.references),
				value,
				[
					...STANDARD_EXCLUDED_PROPERTIES,
					'references'
				]
			);
		}],
		['Error/Application', ApplicationError, value => {
			return assignExtraProperties(
				new ApplicationError(value.message),
				value
			);
		}],
		['Error/Internal', InternalError, value => {
			return assignExtraProperties(
				new InternalError(value.message),
				value
			);
		}],
		['Error/RemoteServer', RemoteServerError, value => {
			return assignExtraProperties(
				new RemoteServerError(value.message),
				value
			);
		}],
		['Error/Timeout', TimeoutError, value => {
			return assignExtraProperties(
				new TimeoutError(value.message, value.timeout),
				value
			);
		}],
		['Error/LimitExceeded', TimeoutError, value => {
			return assignExtraProperties(
				new TimeoutError(value.message, value.timeout),
				value
			);
		}]
	]);
}

function setup(normalizer: DataNormalizer, defs: Definition[]) {
	for (const [name, clazz, denormalizer] of defs) {
		normalizer.registerNormalization({
			name,
			clazz,
			normalizer: normalizeError,
			denormalizer
		});
	}
}

type Definition = [string, any, ((value: any) => any) | undefined];

export function setupSerializerForStandardErrors(serializer: Serializer<any>) {
	setup(serializer.normalizer, [
		['Error', global.Error, value => {
			return new Error(value.message);
		}],
		['TypeError', global.TypeError, value => {
			return new TypeError(value.message);
		}],
		['ReferenceError', global.ReferenceError, value => {
			return new ReferenceError(value.message);
		}],
	]);
}
