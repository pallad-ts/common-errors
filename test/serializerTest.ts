import {JSONAdapter, Serializer, DataNormalizer} from "alpha-serializer";
import {setupSerializer, setupSerializerForStandardErrors} from "@src/serializer";
import {InternalError} from "@src/InternalError";
import {NotFoundError} from "@src/NotFoundError";
import {ApplicationError} from "@src/ApplicationError";
import {RemoteServerError} from "@src/RemoteServerError";

describe('serializer', () => {

    let serializer: Serializer;

    beforeEach(() => {
        serializer = new Serializer(new JSONAdapter(), new DataNormalizer());
    });

    describe('common errors', () => {
        beforeEach(() => {
            setupSerializer(serializer);
        });

        it('InternalError', () => {
            const error = new InternalError('Some internal error', new Error('foo bar'));

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });

        it('NotFoundError', () => {
            const error = new NotFoundError('message', {id: 10, type: 'test'});

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });

        it('ApplicationError', () => {
            const error = new ApplicationError('message');

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });

        it('ApplicationError', () => {
            const error = new RemoteServerError('message');

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });
    });

    describe('standard errors', () => {
        beforeEach(() => {
            setupSerializerForStandardErrors(serializer);
        });

        it('TypeError', () => {
            const error = new TypeError('Hello');

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });

        it('Error', () => {
            const error = new Error('Hello');

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });

        it('ReferenceError', () => {
            const error = new ReferenceError('Hello');

            expect(serializer.normalizer.normalize(error))
                .toMatchSnapshot();
        });
    });
});
