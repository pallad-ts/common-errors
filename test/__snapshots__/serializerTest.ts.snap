// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializer common errors ApplicationError 1`] = `
Object {
  "@type": "Error/RemoteServer",
  "value": Object {
    "message": "message",
    "name": "RemoteServerError",
  },
}
`;

exports[`serializer common errors ApplicationError standard 1`] = `
Object {
  "@type": "Error/Application",
  "value": Object {
    "message": "message",
    "name": "ApplicationError",
  },
}
`;

exports[`serializer common errors ApplicationError with code 1`] = `
Object {
  "@type": "Error/Application",
  "value": Object {
    "code": "code",
    "message": "message",
    "name": "ApplicationError",
  },
}
`;

exports[`serializer common errors InternalError 1`] = `
Object {
  "@type": "Error/Internal",
  "value": Object {
    "message": "Some internal error",
    "name": "InternalError",
  },
}
`;

exports[`serializer common errors NotFoundError 1`] = `
Object {
  "@type": "Error/NotFound",
  "value": Object {
    "message": "message",
    "name": "NotFoundError",
    "references": Object {
      "id": 10,
      "type": "test",
    },
  },
}
`;

exports[`serializer standard errors Error 1`] = `
Object {
  "@type": "Error",
  "value": Object {
    "message": "Hello",
    "name": "Error",
  },
}
`;

exports[`serializer standard errors ReferenceError 1`] = `
Object {
  "@type": "ReferenceError",
  "value": Object {
    "message": "Hello",
    "name": "ReferenceError",
  },
}
`;

exports[`serializer standard errors TypeError 1`] = `
Object {
  "@type": "TypeError",
  "value": Object {
    "message": "Hello",
    "name": "TypeError",
  },
}
`;
